### PR TITLE
ci: Run Nutanix provider e2e tests on self-hosted runner

### PR DESCRIPTION
**Pull request by jimmidyson**
Modify e2e workflow to Run Nutanix jobs on self-hosted-github runner

This is necessary as we want some e2e jobs to run on hosted runners
and some to run on self-hosted runners. The actions runner controller
does not support using multiple labels to target the runners
See: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/using-actions-runner-controller-runners-in-a-workflow#using-runner-scale-set-names

Depends on #754.

### CODE REVIEW

File: .github/workflows/e2e.yml [hunk 1]
```diff
@@ -27,41 +31,65 @@ jobs:
...
-          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
-          restore-keys: |
-            ${{ runner.os }}-go-
+          enable-cache: "true"
+          skip-nix-installation: ${{ inputs.provider == 'Nutanix' }}
 
       # The default disk size of these runners is ~14GB, this is not enough to run the e2e tests.
       # Cleanup the disk, see upstream discussion https://github.com/actions/runner-images/issues/2840.
       - name: Cleanup Disk Space
+        if: inputs.runs-on != 'self-hosted-ncn-dind'
```
Comment (jimmidyson) [inline .github/workflows/e2e.yml:48]: Why make this conditional? Just clean it up anyway, it doesn't take any time, especially compared to the other steps.

Comment (thunderboltsid) [reply .github/workflows/e2e.yml:48]: because the reason for the cleanup is lack of disk space in hosted  runners which is not something we're constrained by here.

File: hack/tools/redact-artifacts/main.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		}
+
+		var unstructured map[string]interface{}
+		if err := yaml.Unmarshal([]byte(doc), &unstructured); err != nil {
+			fmt.Printf("Failed to unmarshal document in file %s: %v\n", path, err)
+			continue
+		}
+
+		if kind, ok := unstructured[kindKey].(string); ok && kind == secretKind {
+			redactSecret(unstructured)
```
Comment (jimmidyson) [inline hack/tools/redact-artifacts/main.go:71]: Make this return an error as well.

File: hack/tools/redact-artifacts/main.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		if err != nil {
+			fmt.Printf("Failed to marshal document in file %s: %v\n", path, err)
+			continue
+		}
+
+		output = append(output, string(redacted))
+	}
+
+	if err := os.WriteFile(path, []byte(strings.Join(output, "\n"+documentSeparator+"\n")), 0o600); err != nil {
+		fmt.Printf("Failed to write file %s: %v\n", path, err)
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply hack/tools/redact-artifacts/main.go:84]: as discussed offline: omitting upload and redaction altogether for now.

File: api/v1alpha1/zz_generated.deepcopy.go [hunk 4]
```diff
@@ -1294,7 +1294,9 @@ func (in *NutanixNodeSpec) DeepCopy() *NutanixNodeSpec {
 }
 
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
-func (in *NutanixPrismCentralEndpointCredentials) DeepCopyInto(out *NutanixPrismCentralEndpointCredentials) {
+func (in *NutanixPrismCentralEndpointCredentials) DeepCopyInto(
+	out *NutanixPrismCentralEndpointCredentials,
+) {
```
Comment (jimmidyson) [inline api/v1alpha1/zz_generated.deepcopy.go:1299]: This is unrelated - please revert.

File: examples/capi-quick-start/nutanix-cluster-calico-crs.yaml [hunk 5]
```diff
@@ -67,7 +67,7 @@ spec:
             credentials:
               secretRef:
                 name: ${CLUSTER_NAME}-pc-creds
-            strategy: HelmAddon
+            strategy: ClusterResourceSet
```
Comment (jimmidyson) [inline examples/capi-quick-start/nutanix-cluster-calico-crs.yaml:70]: Please revert all these example cluster changes and rebase on main now https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/765 is merged.

File: hack/examples/overlays/clusters/nutanix/calico/crs/kustomization.yaml.tmpl [hunk 6]
```diff
@@ -17,3 +17,6 @@ patches:
   - target:
       kind: Cluster
     path: ../../../../../patches/crs-strategy.yaml
+  - target:
+      kind: Cluster
+    path: ../../../../../patches/nutanix/ccm-crs-strategy.yaml
```
Comment (jimmidyson) [inline hack/examples/overlays/clusters/nutanix/calico/crs/kustomization.yaml.tmpl:22]: Revert all these changes please.

File: test/e2e/config/caren.yaml [hunk 7]
```diff
@@ -8,6 +8,14 @@ images:
...
   - name: ghcr.io/nutanix-cloud-native/caren-helm-reg:${E2E_IMAGE_TAG}-${GOARCH}
     loadBehavior: mustLoad
+  - name: docker.io/mesosphere/cluster-api-controller:${CAPI_VERSION}-d2iq.0
+    loadBehavior: mustLoad
+  - name: docker.io/mesosphere/kubeadm-bootstrap-controller:${CAPI_VERSION}-d2iq.0
+    loadBehavior: mustLoad
+  - name: docker.io/mesosphere/kubeadm-control-plane-controller:${CAPI_VERSION}-d2iq.0
+    loadBehavior: mustLoad
+  - name: docker.io/mesosphere/capd-manager:${CAPD_VERSION}-d2iq.0
+    loadBehavior: mustLoad
```
Comment (jimmidyson) [inline test/e2e/config/caren.yaml:18]: Why do we need to do this?

File: test/e2e/config/caren.yaml [hunk 8]
```diff
@@ -188,15 +196,15 @@ variables:
   # IP/FQDN of Prism Central.
   # NOTE: This has to be overridden by specifying the env var NUTANIX_ENDPOINT when running the e2e tests. It is
   # set as empty here to enable running the e2e tests for non-nutanix providers locally without setting the env var.
-  NUTANIX_ENDPOINT: ""
+  #  NUTANIX_ENDPOINT: ""
```
Comment (jimmidyson) [inline test/e2e/config/caren.yaml:199]: Please revert - this makes it easier to run tests locally and shouldn't be required for CI.

File: api/v1alpha1/zz_generated.deepcopy.go (hunk 4 above)
Comment (thunderboltsid) [reply api/v1alpha1/zz_generated.deepcopy.go:1299]: This came from `make pre-commit`

File: test/e2e/config/caren.yaml (hunk 7 above)
Comment (thunderboltsid) [reply test/e2e/config/caren.yaml:18]: Because the kind cluster can hit rate limits

File: api/v1alpha1/zz_generated.deepcopy.go (hunk 4 above)
Comment (jimmidyson) [reply api/v1alpha1/zz_generated.deepcopy.go:1299]: Yes I know but it's unrelated and will not fail in CI - need to figure out how to get golines to ignore these generated files...

File: .github/workflows/e2e.yml [hunk 9]
```diff
@@ -27,41 +31,77 @@ jobs:
...
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+
       - name: Run e2e tests
-        run: devbox run -- make e2e-test E2E_LABEL='provider:${{ inputs.provider }}' E2E_SKIP='${{ inputs.skip }}' E2E_FOCUS='${{ inputs.focus }}'
+        run: devbox run -- make e2e-test E2E_LABEL='provider:${{ inputs.provider }} && cni:Cilium && addonStrategy:HelmAddon' E2E_SKIP='${{ inputs.skip }}' E2E_FOCUS='${{ inputs.focus }}'
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply .github/workflows/e2e.yml:87]: This was due to a discussion with @dkoshkin where we wanted to start out with just cilium and helmaddon as that's what we intend to originally support.

File: test/e2e/config/caren.yaml (hunk 7 above)
Comment (jimmidyson) [reply test/e2e/config/caren.yaml:18]: These should really come from `ghcr.io` as they are open source projects. I'll try to get that fixed and then we will be able to remove this.

File: .github/workflows/e2e.yml [hunk 10]
```diff
@@ -27,10 +39,16 @@ jobs:
         with:
           fetch-depth: 0
 
+      - uses: cachix/install-nix-action@V27
```
Comment (dkoshkin) [inline .github/workflows/e2e.yml:42]: Please add a `name` and some comments on why this is different depending on the runner.

File: test/e2e/self_hosted_test.go [hunk 11]
```diff
@@ -22,103 +22,103 @@ import (
...
-											flavour,
-											"wait-helmrelease",
-										),
-										ClusterResourceSetIntervals: e2eConfig.GetIntervals(
-											flavour,
-											"wait-clusterresourceset",
-										),
-									},
-								)
+	for _, provider := range []string{"Docker", "Nutanix"} {
```
Comment (dkoshkin) [inline test/e2e/self_hosted_test.go:25]: The diff is hard to read, any other changes besides adding the provider?

**Review by dkoshkin — APPROVED** [review]

File: test/e2e/self_hosted_test.go (hunk 11 above)
Comment (jimmidyson) [reply test/e2e/self_hosted_test.go:25]: Yes just adding the provider here.

Much easier if you switch to side by side and ignore whitespace btw...

image

File: .github/workflows/checks.yml [hunk 12]
```diff
@@ -59,14 +59,23 @@ jobs:
...
+          - Nutanix
           # Uncomment below once we have the ability to run e2e tests on other providers from GHA.
           # - AWS
-          # - Nutanix
+        cni:
+          - Calico
+          - Cilium
+        addon-strategy:
+          - ClusterResourceSet
+          - HelmAddon
```
Comment (jimmidyson) [inline .github/workflows/checks.yml:70]: I like this splitting up! We'll need to update the required checks in canaveral to allow this PR to be merged though.

Comment (jimmidyson) [reply .github/workflows/checks.yml:70]: Actually I wonder if we can just make this e2e labels instead of splitting up the provider/cni/addon-strategy like this.

We could continue to run Docker e2e as a single build in that case, and split up the longer running builds like AWS, Nutanix, etc as we need to - more flexibility. WDYT?

Comment (thunderboltsid) [reply .github/workflows/checks.yml:70]: That could work as well but then we'd have to extract the provider and the runs-on value for workflow conditionals from the labels in GitHub workflow steps which could get ugly.

Comment (thunderboltsid) [reply .github/workflows/checks.yml:70]: as discussed offline, we won't proceed with this.

File: test/e2e/config/caren.yaml [hunk 13]
```diff
@@ -194,9 +202,9 @@ variables:
   # # Disable Prism Central certificate checking. Default: false
   # NUTANIX_INSECURE: false
   # # Prism Central user
-  NUTANIX_USER: ""
+  #  NUTANIX_USER: ""
```
Comment (jimmidyson) [inline test/e2e/config/caren.yaml:205]: Please revert.

File: test/e2e/config/caren.yaml [hunk 14]
```diff
@@ -194,9 +202,9 @@ variables:
...
   # # Prism Central password
-  NUTANIX_PASSWORD: ""
+  #  NUTANIX_PASSWORD: ""
```
Comment (jimmidyson) [inline test/e2e/config/caren.yaml:207]: Please revert.

File: test/e2e/config/caren.yaml (hunk 13 above)
Comment (thunderboltsid) [reply test/e2e/config/caren.yaml]: as discussed offline, we should fix this in CAPX infrastructure-components.yaml. This should not be a requirement for clusterctl init.

**Review by mhrabovcin — APPROVED** [review]
